tiledb/sm/filesystem/ls_scanner.h

```diff
@@ -0,0 +1,232 @@
+
+using DirectoryFilter = std::function<bool(const std::string_view&)>;
+/** Static DirectoryFilter used as default argument. */
+[[maybe_unused]] static bool accept_all_dirs(const std::string_view&) {
+  return true;
+}
+
+/** Type defintion for objects returned from ls_recursive. */
+using LsObjects = std::vector<std::pair<std::string, uint64_t>>;
+
+/**
+ * LsScanIterator iterates over the results of ls requests wrapped by classes
+ * deriving from LsScanner. See S3Scanner as an example.
+ *
+ * @tparam T The LsScanner type that created this iterator.
+ * @tparam U The data type stored by this iterator.
+ */
+template <class T, class U>
+class LsScanIterator {
+ public:
+  using value_type = U;
+  using difference_type = ptrdiff_t;
+  using pointer = const U*;
+  using reference = const U&;
+  using iterator_category = std::input_iterator_tag;
+
+  /** Default constructor. */
+  LsScanIterator() noexcept = default;
+
+  /**
+   * Constructor.
+   *
+   * @param scanner The scanner that created this iterator.
+   */
+  explicit LsScanIterator(T* scanner) noexcept
+      : scanner_(scanner)
+      , ptr_(scanner->begin_) {
+  }
+
+  /**
+   * Constructor.
+   *
+   * @param scanner The scanner that created this iterator.
+   * @param ptr Pointer to set as the current object.
+   */
+  LsScanIterator(T* scanner, pointer ptr) noexcept
+      : scanner_(scanner)
+      , ptr_(ptr) {
+  }
+
+  /** Copy constructor. */
+  LsScanIterator(const LsScanIterator& rhs) noexcept {
+    ptr_ = rhs.ptr_;
+    scanner_ = rhs.scanner_;
+  }
+
+  /** Copy assignment operator. */
+  LsScanIterator& operator=(LsScanIterator rhs) noexcept {
+    if (&rhs != this) {
+      std::swap(ptr_, rhs.ptr_);
+      std::swap(scanner_, rhs.scanner_);
+    }
+    return *this;
+  }
+
+  /**
+   * Dereference operator.
+   *
+   * @return The current object being visited.
+   */
+  constexpr reference operator*() const noexcept {
+    return *ptr_;
+  }
+
+  /**
+   * Dereference operator.
+   *
+   * @return The current object being visited.
+   */
+  constexpr pointer operator->() const noexcept {
+    return ptr_;
+  }
+
+  /**
+   * Prefix increment operator.
+   * Calls the scanner's next() method to advance to the next object.
+   *
+   * @return Reference to this iterator after advancing to the next object.
+   */
+  LsScanIterator& operator++() {
+    scanner_->next(ptr_);
+    return *this;
+  }
+
+  /**
+   * Postfix increment operator.
+   * Calls next() method to advance to the next object via prefix operator.
+   *
+   * @return Copy of this iterator prior to advancing to the next object.
+   */
+  LsScanIterator operator++(int) {
+    LsScanIterator tmp(*this);
+    operator++();
+    return tmp;
+  }
+
+  /**
+   * @return Iterator to the beginning of the results being iterated on.
+   *    Input iterators are single-pass, so we return a copy of this iterator at
+   *    it's current position.
+   */
+  LsScanIterator begin() {
```

Comment:
I don't think there's a reason to implement this function, or `end()` below either. It's not meaningful for an input iterator.